connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/LoaderSwap.java

```diff
@@ -24,13 +24,32 @@ public class LoaderSwap implements AutoCloseable {
 
     private final ClassLoader savedLoader;
 
-    public LoaderSwap(ClassLoader savedLoader) {
+    public static LoaderSwap use(ClassLoader loader) {
+        ClassLoader savedLoader = compareAndSwapLoaders(loader);
+        try {
+            return new LoaderSwap(savedLoader);
+        } catch (Throwable t) {
+            compareAndSwapLoaders(savedLoader);
+            throw t;
+        }
+    }
```

Comment:
This is not re-introducing the static logic, it is just refactoring to eliminate the open-ended Plugins.compareAndSwap* methods.

This method is only called in two places: by DelegatingClassLoader.scanPluginPath (before scanning is finished) and Plugins.withClassLoader (after scanning is finished).

I've dropped the visibility and made the DCL call-site mock-able.

Reply:
This does still involve static logic for classloader swapping, though. And the comment about internal use doesn't seem very helpful since the way we use that term ("internal") has to do with public vs. private API; it's not really clear to people that (or why) they shouldn't just upgrade the visibility to public.

Ultimately I'd prefer to see this logic duplicated in two places (`DelegatingClassLoader::withClassLoader` and `Plugins::withCLassLoader`) rather than introduce a new API that might be misused in the future.